### DISCUSSION TITLE
Trying to format a time and also a date, help requested

### DISCUSSION THREAD

**Discussion by graydav1** [WerWolv/ImHex#1373] [Pattern Language Q&A] [answered]
I'm writing a pattern for some data I've got (serial data being sent to some hardware), and I'm trying to parse three bytes as a time (H:M:S). I've read the documentation, but there isn't a whole lot of examples.

The data is simply the hour, minute and second as integers. So for example, 08:32:00 would be represented in the data as `08 20 00`

In addition to that, I've got two bytes that make up a date. I'd like to parse that as a date (assuming the year will be the current year). I've looked over the docs, but there's no examples of formatting dates, nor assuming data (e.g. just getting the current year)

I've looked through the discussions, the documentation, and also checked out the included patterns, but I can't seem to work it out.

**Comment by WerWolv** [comment, marked as answer]
Hey, you don't really need to mess with any date-time things since the format is super easy. The stuff that's included in the standard library deals with Unix / Epoch time.

A pattern for this could maybe look like this:
```cpp
struct DateTime {
    u8 hour, minute, second;
    u8 day, month;
} [[format("format_date_time")]];

fn format_date_time(ref auto dateTime) {
    return std::format("{}:{}:{} {}.{}", dateTime.hour, dateTime.minute, dateTime.second, dateTime.day, dateTime.month);
};
```

**Reply by graydav1** [reply]
Ah, I see how it all works now. Makes perfect sense, and that works great. I added `{:02}` to pad out the zeroes after realizing this works like C++'s format function and my data looks great.

Thanks!

**Reply by jumanji144** [reply]
For future reference, the format is just using the `fmt` libraries formating, so it is indeed using the `std::format` format

**Reply by WerWolv** [reply]
Great, you're welcome!
Yeah we use libfmt for the underlying formatting which is what C++'s `std::format` is based on. You can find the full documentation here: https://fmt.dev/latest/syntax.html
Basically everything is supported
